source/common/ssl/context_impl.cc

```diff
@@ -365,11 +379,120 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
                                   const unsigned char* in, unsigned int inlen, void* arg) -> int {
-                                 return static_cast<ServerContextImpl*>(arg)->alpnSelectCallback(
-                                     out, outlen, in, inlen);
+                                 return static_cast<ServerContextImpl*>(arg)
+                                     ->alpnSelectCallback(out, outlen, in, inlen);
                                },
                                this);
   }
+
+  if (!config.sessionTicketKeys().empty()) {
+    session_ticket_keys_.resize(config.sessionTicketKeys().size());
+    for (unsigned i = 0; i < config.sessionTicketKeys().size(); ++i) {
+      // If this changes, need to figure out how to deal with key files
+      // that previously worked.  For now, just assert so we'll notice that
+      // it changed if it does.
+      static_assert(sizeof(SessionTicketKey) == 80, "Input is expected to be this size");
+
+      const std::vector<uint8_t>& src_key = config.sessionTicketKeys().at(i);
+      SessionTicketKey& dst_key = session_ticket_keys_.at(i);
+
+      if (src_key.size() != sizeof(SessionTicketKey)) {
+        throw EnvoyException(fmt::format("Incorrect TLS session ticket key length.  "
+                                         "Index {}, length {}, expected length {}.",
+                                         i, src_key.size(), sizeof(SessionTicketKey)));
+      }
+
+      std::copy_n(src_key.begin(), dst_key.name.size(), dst_key.name.begin());
+      size_t pos = dst_key.name.size();
+      std::copy_n(src_key.begin() + pos, dst_key.aes_key.size(), dst_key.aes_key.begin());
+      pos += dst_key.aes_key.size();
+      std::copy_n(src_key.begin() + pos, dst_key.hmac_key.size(), dst_key.hmac_key.begin());
+      pos += dst_key.hmac_key.size();
+      ASSERT(src_key.begin() + pos == src_key.end());
+    }
+    SSL_CTX_set_tlsext_ticket_key_cb(ctx_.get(), [](SSL* ssl, uint8_t* key_name, uint8_t* iv,
+                                                    EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx,
+                                                    int encrypt) -> int {
+      ContextImpl* context_impl =
+          static_cast<ContextImpl*>(SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), ssl_context_index_));
+      return dynamic_cast<ServerContextImpl*>(context_impl)
+          ->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
+    });
+  }
+
+  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
+  unsigned session_context_len = 0;
+  if (ca_cert_ != nullptr) {
+    // If there is a CA cert, only allow resuming in a context with the same
+    // CA cert.  This ensures that the client is always validated against
+    // the correct CA cert, even if session resumption across different listeners
+    // is enabled.
+    int ret = X509_digest(ca_cert_.get(), EVP_sha256(), session_context_buf, &session_context_len);
```

Comment:
This is definitely not enough and you should also include all validation options, possibly also listener's name.

With the current code, session established with a listener that has `ca_cert_file` configured can be used to resume session on any listener that has the same `ca_cert_file`, even when extra restrictions are in place (e.g.  `verify_subject_alt_name`), effectively bypassing them.